### ISSUE TITLE
Nuxt Layers - array duplicity within app.config.ts (

### ISSUE THREAD

**Issue by kleinpetr** [nuxt/nuxt#21287]
### Environment

```
------------------------------
- Operating System: Linux
- Node Version:     v16.20.0
- Nuxt Version:     3.5.2
- Nitro Version:    2.4.1
- Package Manager:  unknown
- Builder:          vite
- User Config:      -
- Runtime Modules:  -
- Build Modules:    -
------------------------------
```

### Reproduction

https://stackblitz.com/github/kleinpetr/nuxt-layers-configs

Stackblitz uses pnpm older pnpm, so the `install` fails, but feel free to clone it, or just go through the code, it's very minimalist 

### Describe the bug

I am facing 2 issues when using this architecture

![image](https://github.com/nuxt/nuxt/assets/17269309/063dc89d-a946-4889-b369-c9a04279d635)

### Issue 1:
When I use array in the app.config for the base layer, at the end it multiplies values, as you can see in the devtools

![image](https://github.com/nuxt/nuxt/assets/17269309/9131208d-9726-487e-b67c-ebbe93545fae)

### Issue 2:
When I use `useHead` within `nuxt.config.ts` it also duplicates it, as you can see in the generated source code

![image](https://github.com/nuxt/nuxt/assets/17269309/7d4938ae-ff41-4b0b-a264-d5796c4e0dc5)

### Additional context

I am not sure if this is not the expected behavior of `defu` so let me know ;)

**Comment by wilsonpinto** [comment]
@kleinpetr for issue 1, this is related with https://github.com/nuxt/nuxt/issues/15649, currently it's used [Function Merger](https://github.com/unjs/defu#function-merger) for arrays.

Looks like to fix this you need to redefine the defineAppConfig in "App":

```
 baseLayer: {
    values: (val) => [val]
  }
  ```

**Comment by pi0** [comment]
Yes, we use function merger. I think improving docs would be a nice idea..

**Comment by kleinpetr** [comment]
> @kleinpetr for issue 1, this is related with #15649, currently it's used [Function Merger](https://github.com/unjs/defu#function-merger) for arrays.
> 
> Looks like to fix this, you need to redefine the `values` in the defineAppConfig (App):
> 
> ```js
>  baseLayer: {
>     values: (val) => [val]
>   }
> ```

but it's cannot be done with custom types..

**Comment by JoelForamitti** [comment]
If there are more complex dependencies between layers like in the image below, each adding values to the same array in app.config, is there a good way to merge an array without duplicates?

![image](https://github.com/nuxt/nuxt/assets/57440945/f8c9b53e-b769-4ee1-a69b-496740e4cc78)

**Comment by kleinpetr** [comment]
yes that you can do for your custom configs.. but it cannot be used for 3d party packages such as i18n or cookie-control in that case the values are duplicated.. and if I try to use @wilsonpinto solution it is not working properly..

![image](https://github.com/nuxt/nuxt/assets/17269309/c3d7c8d5-cc18-48d6-8c06-67eecc80762d)

So is there any other way how I could define, that I want to replace values instead of merge them?

**Comment by derHodrig** [comment]
This type issue `() => [] not assignable to []` is a huge blocker.

**Comment by danielroe** [comment]
@pi0 To fix this, I think we'd need to change this in untyped, instead, right?

Or were you thinking we need to (separately) improve this in defu?

**Comment by pi0** [comment]
Sure also for untyped (does defuFn already properly resolves function types btw? i guess it doesn't)

**Comment by derHodrig** [comment]
I dont know if this is related, but if I have the following index.d.ts in my layer

```
import { AppConfigInterface } from '#imports'

declare module 'nuxt/schema' {
  interface AppConfigInput extends AppConfigInterface {}
}
declare module 'nuxt/schema' {
  interface AppConfig extends AppConfigInterface {}
}

// It is always important to ensure you import/export something when augmenting a type
export {}
```
The appConfig output types are not set to AppConfigInterface
eg. if I have a Menu Array and those have no childItems, the auto generated type of appConfig will not know of childItems: MenuItem[] but its given in the AppConfigInterface.

https://nuxt.com/docs/guide/directory-structure/app-config#typing-app-config-output

But maybe this is a different issue or misconfigured in my layer.

**Comment by kleinpetr** [comment]
@derHodrig yes I am facing the same issue, so in my opinion declaring the app config schema is quite useless since it s generated from the values itself, so for example when I want to avoid some value I have to do something like this

```export default defineAppConfig({
  develit3: {
    supportedTokens: undefined as (SupportedTokens | undefined),
  },
})
```

**Comment by Flo0806** [comment]
@danielroe Can also be closed - all is by design and documented.

https://nuxt.com/docs/4.x/directory-structure/app/app-config#merging-strategy
https://github.com/unjs/defu#function-merger

**Comment by danielroe** [comment]
nuxtathon closed @Flo0806
